proposals/break-continue-in-inline-lambdas.md

```diff
@@ -0,0 +1,175 @@
+body of `try-catch`), it would be only logical to allow the same for lambdas passed as parameters to inline functions. At the
+moment only non-local `return`s are supported. That's why we are proposing to add support for non-local `break`/`continue` as
+well.
+
+So the Motivation is the following:
+
+1. Make inline lambdas truly first-class language constructs that are seamless to use.
+2. Fix consistency with non-local `return`. It's inconsistent that currently we allow non-local `return` statements but don't
+   allow non-local `break` and `continue` statements.
+
+## Design problems / Alternative approach
+
+The only design problem we managed to find so far is nesting of "loop-like" inline functions into `for`/`while` loops.
+
+```kotlin
+for (i in 0 until 100) {
+    (0 until 200).forEach {
+        if (it == 42) break // break `for` loop, but new users may think that break from `forEach` loop
+    }
+}
+```
+
+The alternative approach may be to make `break`/`continue` apply to the closest `for`, `while` or "loop-like" functions. We will
+never allow unlabeled `break`/`continue` for "loop-like" functions. The rationale is simple: if Kotlin allowed unlabeled
+`break`/`continue` for "loop-like" functions then it would mean that for such code
+
+```kotlin
+while (cond) {
+    someInlineFunctionThatAcceptsLambda {
+        break
+    }
+}
+```
+
+users wouldn't be able to understand whether `break` breaks from `someInlineFunctionThatAcceptsLambda` or from `while` without
+looking into `someInlineFunctionThatAcceptsLambda` implementation. It complicates reasoning about `break` and `continue`
+statements. In other words, it's hard to distinguish "loop-like" functions from other lambdas if the call-site doesn't differ
+syntactically.
+
+Also, right now we have such an invariant in Kotlin: a pair of **keywords** `break`/`continue` tied to the closest `while` or
+`for` **keywords**. `return` **keyword** is tied to the closest `fun` **keyword**. The proposed alternative solution breaks this
+invariant.
+
+* **Q**: Why not make labels mandatory for non-local `break`/`continue`? That would eliminate `forEach` puzzler.
+* **A**: Because Kotlin doesn't require labels for non-local `return`. For sake of consistency, We should either require labels
+  for both `break`/`continue` and `return` cases or don't require the labels again for the both cases.
+
+* **Q**: Ok, why not make labels mandatory for both non-local `return` and non-local `break`/`continue` then?
+* **A**: We can't make non-local `return` to be obligatory labeled because it would break inline functions first-class
+  citizenship in language constructs. E.g. we value "return if not null" idiomatic Kotlin construct `nullable?.let { return it }`.
+  If non-local `return` were obligatory labeled then the code would be much more verbose
+  `nullable?.let { return@functionName it }`.
```

Comment:
I have a third question for which I don't have an answer:

* **Q**: Why not make labels mandatory for both non-local `return` and non-local `break`/`continue` but only for lambdas that don't have `callsInPlace(block, InvocationKind.EXACTLY_ONCE)` contract?

Reply:
@nikitabobko We cannot make labels mandatory for non-local returns as that would break existing code. Instead, we have a simple rule: an unlabeled return within a lambda is always non-local, a labeled one, well, returns to the label.

Reply:
I think we've discussed this during design. We cannot make it an error, because we don't have a concept of loop-like functions in the language. We simply use a heuristic that is based on AT_MOST_ONCE/EXACTLY_ONCE contract. We know that functions with such a contract are _definitely not loop-like_, but the converse is not true. In reality, we don't know much about lambdas not marked and such, so a hard-requirement (as in "produces compiler ERROR") to use labelled return there would be an overkill. It much better belongs to a category of inspections, a kind of style issue.

Reply:
@elizarov Thanks! Add a section describing this decision https://github.com/Kotlin/KEEP/pull/327/commits/cb55a237387718a45d38edd5169e43c787d7f3cc